### PR TITLE
coding guidelines: comply with MISRA C:2012 Rule 13.3

### PR DESCRIPTION
**Pull request by Abramo-Bagnara**
In particular:

- avoided ++/-- on volatile

- avoided to have more than one ++/-- in for third clause

- moved ++/-- before or after the value use

- avoided pointless ++/--

- replaced consecutive ++ with a single +=

Signed-off-by: Abramo Bagnara <abramo.bagnara@bugseng.com>

### CODE REVIEW

File: kernel/mempool.c [hunk 1]
```diff
@@ -33,7 +33,7 @@ static void *z_heap_aligned_alloc(struct k_heap *heap, size_t align, size_t size
 
 	heap_ref = mem;
 	*heap_ref = heap;
-	mem = ++heap_ref;
+	mem = heap_ref + 1;
```
Comment (dcpleung) [inline kernel/mempool.c:36]: `heap_ref` still needs to be incremented.

Comment (Abramo-Bagnara) [reply kernel/mempool.c:36]: I don't see any use of local variable heap_ref after this point... Am I missing something?

**Review by andyross — COMMENTED** [review]
This touches a lot of subsystems in a single patch, often in code that looks edge-casey and poorly exercised in CI (in particular I know for sure there's no unit test anywhere of the EFI console output).

If one of these turns out to regress something that we only notice a few months in the future, this is going to be a very hard patch to revert to test, given how many lines it is.  Suggest splitting the changes out into separate patches by subsystem/driver.

**Comment by npitre** [comment]
I have no problem with the post-indexing removal. I can see the point
of such change. I think you should split it out into a PR of its own
so that the least controvertial changes could go in faster.

BTW there is that hunk:

```
-               while (*--buf == '0') {
-                       ;
+               --buf;
+               while (*buf == '0') {
+                       --buf;
                }
```

Can't you turn that into the following instead:

```
                do {
                    --buf;
                } while (*buf == '0');
```

That's clearer and shorter too.

That being said ...

I do strongly disagree with qualifying small literals. Referring to a
conversation we had a while ago:

```
void foo(my_type_t x) { if (x < 5U) { bar(x); } }
```

Here things will go wrong the moment `my_type_t` is signed and `x` is
negative. Whereas:

```
void foo(my_type_t x) { if (x < 5) { bar(x); } }
```

Here you'll get the proper behavior all the time whether `my_type_t`
is signed or not. The principle of least surprise is respected.

So I stand by my statement that qualified small literals is evil and
MISRA is wrong on that one. This overzealous qualifying of small literals
must stop. It should *not* be encouraged _unless_ it is warranted by an
explicit purpose like bitmask operations where shifting the sign bit
would create issues. Otherwise it is just useless and ugly.

This goes for arithmetic operations too.

And equality tests against a qualified zero are just plain stupid, period.

So... to recap, I'm fine with the top commit. Not so much with the
other ones.

**Comment by Abramo-Bagnara** [comment]
> Can't you turn that into the following instead:
> 
> ```
>                 do {
>                     --buf;
>                 } while (*buf == '0');
> ```
> 
> That's clearer and shorter too.

I agree, changed now.

**Comment by Abramo-Bagnara** [comment]
> This touches a lot of subsystems in a single patch, often in code that looks edge-casey and poorly exercised in CI (in particular I know for sure there's no unit test anywhere of the EFI console output).
> 
> If one of these turns out to regress something that we only notice a few months in the future, this is going to be a very hard patch to revert to test, given how many lines it is. Suggest splitting the changes out into separate patches by subsystem/driver.

I don't understand what you write: the commit touches only 14 files and none of them is under subsys or drivers folders.

**Comment by ceolin** [comment]
I think the pr should be split in more commits with more information about the changes. Also some changes make the code more bureaucratic to read, but in general the changes look good to me.

**Review by ceolin — APPROVED** [review]

**Pull request merged**